README.org
Fllow this repository to be a electronic engineer (EE).


RadarWarn/history.org
# -*- eval: (setq org-download-image-dir (file-name-sans-extension (buffer-name))); -*-
# -*- org-export-babel-evaluate: nil; -*-
#+HTML_HEAD: <link rel="stylesheet" type="text/css" href="../orgstyle.css"/>
#+OPTIONS: ':nil *:t -:t ::t <:t H:3 \n:t arch:headline author:t c:nil S:nil -:nil
#+OPTIONS: creator:nil d:(not "En") date:t e:t email:nil f:t inline:t
#+OPTIONS: num:t p:nil pri:nil prop:nil stat:t tags:t tasks:t tex:t timestamp:t
#+OPTIONS: title:t toc:t todo:t |:t 
#+OPTIONS: ^:{}
#+LATEX_CLASS: ctexart
#+STARTUP: entitiespretty:t
#+TITLE: 雷达告警的那些事
#+SELECT_TAGS: export
#+EXCLUDE_TAGS: noexport

THE RADAR WARNING STORY




:En:
IN THE BEGINNING YEAR: 1965
In July 1965, an American F-4 Phantom II was shot down over North Vietnam. This loss represented the first American aircraft to be downed by an SA-2 Surface-to-Air Missile (SAM) in the Southeast Asian conflict. This event heralded the beginning of a serious and lethal new challenge to American air superiority.
:END:
* 1965:伊始
1965年7月，一架美国F-4“幻影II”在北越被击落。这是第一架在东南亚冲突中被SA-2地对空导弹（SAM）击落的美国飞机。这一事件预示着一场针对美国空中优势的严重的致命新挑战拉开序幕。

:En:
In the succeeding months of the war, American aircraft would attempt to penetrate major targets in the North, and they would fly their missions into the teeth of the most sophisticated air defense network ever deployed in history.
The North Vietnam air defense system, as it evolved in the mid to late 1960s, consisted of AntiAircraft Artillery (AAA), SAMs and MIG fighter aircraft, all closely coordinated through the use of communications and radar. The need to counter this lethal new threat was painfully obvious as losses of aircraft and crew grew to unacceptable levels.
Brigadier General K.C. Dempster,
then Director of Operational Requirements and Development Plans, U.S. Air Force Headquarters, was selected to head a Blue Ribbon Air Staff Task Force to investigate methods of countering these lethal weapon systems in an effort to provide warning and protection to U.S. air crews against this new threat.
The need for aircraft self-protection assets was the driving force behind a seminar held in August 1965, and led to a recommendation for the development of aircraft avionics systems capable of detecting and destroying the controlling radars in these defense networks. The concept was appropriately named Wild Weasel, an animal known to hunt vermin.
Almost in disastrous parallel, a Navy
RA-5C reconnaissance aircraft was lost over Dienbienphu. Again, the culprit was the SA-2.
A $5.5 million aircraft was lost, causing considerable comment that the aircraft was too expensive to introduce into combat. Soviet-built SAM technology was threatening to destroy a much needed intelligence capability that had been developed and deployed at a cost of several hundred million dollars.
In parallel with the Dempster panel, DDR&E convened a committee to evaluate options and make selections for the development of a warning and jamming equipment suite. It was decided that of all the proposals received, a concept called Vector IV offered the best potential solution to the warning problem. As a result of this decision, a then-small company based in Northern California rocketed into national focus. The Vector IV concept was proposed by Applied Technology, Inc.
:END:
在战争的后续几个月中，美国飞机试图渗透北方的主要目标，他们的飞行任务要求飞机必须进入当时史上部署最复杂的防空网络。
北越防空系统是在20世纪60年代中后期发展起来的，由高射炮（AAA）、地空导弹和米格战斗机组成，通过通信和雷达进行密切协调。由于飞机和机组人员的损失增加到不可接受的水平，因此迫切需要应对这种致命的新威胁。
K.C.登普斯特准将，后任美国空军总部的作战需求和发展计划部主任，被选中组建一支蓝丝带空军参谋队伍，负责研究打击这些致命武器系统的方法，以便为美国空勤人员提供警示和保护，有效应对这一新威胁。
飞机自卫能力的需求促成了1965年8月举行的研讨会，会上提出了开发航空电子设备系统用于探测和摧毁防御网雷达的建议。这个概念被命名为野鼬任务（Wild Weasel）。
祸不单行，海军也在越南Dien bien phu损失了一架RA-5C侦察机。罪魁祸首依然是SA-2导弹。损失一架价值550万美元的飞机，这使得多数人都认为这样太昂贵的飞机是无法在战斗中使用的。苏联制造的SAM技术对数百万美元成本的、急需的、高度智能化的作战能力构成了毁灭性威胁。与Dempster小组并行，DDR＆E召集了一个委员会来评估告警和干扰设备组件的指标，并从中做出选择。在收到的所有方案中，一个名为Vector IV的概念最终成为告警问题的最佳潜在解决方案。这一决定也使得一家总部位于北加州的小型公司平步青云，一度成为国家焦点，这家公司就是应用技术公司（Applied Technology, Inc.）。

:En:
In November 1965, the USAF Sacramento Air Material Area awarded Applied Technology a contract for 500 AN/APR-25s, the system nomenclature for the Vector IV concept. The system was based on the use of crystal-video detection techniques in the threat bands only, where the hostile equipment’s relative direction was resolved by a CRT display using vectorial inputs from four antenna patterns. Thus, the initial concept name, Vector IV.
:END:
1965年11月，美国空军萨克拉门托空气材料区授予应用技术公司500批AN/APR-25的生产合同，AN/APR-25是Vector IV概念的系统命名法。该系统仅在威胁的频带中使用晶体视频检波技术，其中敌方信号的到达角通过四个指向不同的天线的矢量输入解决。这也是该系统最初被命名为Vector IV的原因。
[[file:history.org_imgs/20181119_204915_1776-9w.png]]

:En:
Until this time, Applied Technology had been primarily involved in strategic reconnaissance and intelligence technology programs where a large production run consisted of 10 systems. The impact of the AN/APR-25 award was huge. Further, the U.S. Navy awarded Applied Technology a production order to outfit its aircraft with this critical new protective asset.
:END:
在这之前，应用技术公司主要参与战略侦察和情报技术计划，其主要产值由10个系统组成。AN/APR-25带来的利润是如此之大，修正了企业的发展方向。此外，美国海军还向应用技术公司核发了一份生产订单，为其飞机配备这一重要的新式自卫手段。


:En:
Very little was known about the threat in those early days, and the design philosophies of warning systems were based upon gathering as much signal information as available. Later this philosophy would change as the need for sorting analysis and prioritization of lethal pulses became semi-automatic.
:END:
最初，人们对威胁的信息知之甚少，告警系统的设计理念基于尽可能多的收集信号信息。后来，由于需要实现对分类和致命脉冲优先级的半自动化分析，这种理念就慢慢发生了变化。


:En:
On a Friday the 13th in 1966, Applied Technology’s AN/APR-26 radar warning receiver performed successfully in a combat mission and added yet another success for the growing company. In the 1966 – 1967 time frame, Applied Technology was shipping radar homing and warning systems at a rate of 250 per month. Employment levels approached 1,400, a seven-fold increase in just two years.
:END:
1966年的一个星期五，应用技术公司的AN/APR-26雷达告警接收器在一次战斗任务中成功完成了它的首秀，并为这家不断发展的公司带来了又一次成功。在1966至1967年，应用技术公司以每月250个的速度交付雷达制导和预警系统。就业水平接近1,400，仅仅两年就增长了7倍。



gnuradio/zoom_fft.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
##################################################
# GNU Radio Python Flow Graph
# Title: Zoom Fft
# Generated: Tue Aug 29 09:03:08 2017
##################################################

if __name__ == '__main__':
    import ctypes
    import sys
    if sys.platform.startswith('linux'):
        try:
            x11 = ctypes.cdll.LoadLibrary('libX11.so')
            x11.XInitThreads()
        except:
            print "Warning: failed to XInitThreads()"

from gnuradio import blocks
from gnuradio import eng_notation
from gnuradio import filter
from gnuradio import gr
from gnuradio import wxgui
from gnuradio.eng_option import eng_option
from gnuradio.fft import window
from gnuradio.filter import firdes
from gnuradio.wxgui import fftsink2
from gnuradio.wxgui import forms
from grc_gnuradio import wxgui as grc_wxgui
from optparse import OptionParser
import osmosdr
import time
import wx


class zoom_fft(grc_wxgui.top_block_gui):

    def __init__(self):
        grc_wxgui.top_block_gui.__init__(self, title="Zoom Fft")
        _icon_path = "/usr/share/icons/hicolor/32x32/apps/gnuradio-grc.png"
        self.SetIcon(wx.Icon(_icon_path, wx.BITMAP_TYPE_ANY))

        ##################################################
        # Variables
        ##################################################
        self.ratio_chooser = ratio_chooser = 20
        self.zoom_ratio = zoom_ratio = ratio_chooser
        self.samp_rate = samp_rate = 2500000
        self.freq = freq = 94500000

        ##################################################
        # Blocks
        ##################################################
        self.wxgui_fftsink2_0 = fftsink2.fft_sink_c(
        	self.GetWin(),
        	baseband_freq=freq,
        	y_per_div=10,
        	y_divs=10,
        	ref_level=0,
        	ref_scale=2.0,
        	sample_rate=samp_rate/zoom_ratio,
        	fft_size=1024,
        	fft_rate=15,
        	average=False,
        	avg_alpha=None,
        	title='FFT Plot',
        	peak_hold=False,
        	win=window.blackmanharris,
        )
        self.Add(self.wxgui_fftsink2_0.win)
        self._ratio_chooser_chooser = forms.drop_down(
        	parent=self.GetWin(),
        	value=self.ratio_chooser,
        	callback=self.set_ratio_chooser,
        	label='ratio_chooser',
        	choices=[2, 3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20],
        	labels=[],
        )
        self.Add(self._ratio_chooser_chooser)
        self.osmosdr_source_1 = osmosdr.source( args="numchan=" + str(1) + " " + 'airspy' )
        self.osmosdr_source_1.set_sample_rate(samp_rate)
        self.osmosdr_source_1.set_center_freq(freq, 0)
        self.osmosdr_source_1.set_freq_corr(0, 0)
        self.osmosdr_source_1.set_dc_offset_mode(0, 0)
        self.osmosdr_source_1.set_iq_balance_mode(0, 0)
        self.osmosdr_source_1.set_gain_mode(False, 0)
        self.osmosdr_source_1.set_gain(10, 0)
        self.osmosdr_source_1.set_if_gain(20, 0)
        self.osmosdr_source_1.set_bb_gain(20, 0)
        self.osmosdr_source_1.set_antenna('', 0)
        self.osmosdr_source_1.set_bandwidth(0, 0)

        self.low_pass_filter_0 = filter.fir_filter_ccf(zoom_ratio, firdes.low_pass(
        	1, samp_rate, samp_rate/zoom_ratio/2, 100, firdes.WIN_HAMMING, 6.76))
        self.blocks_throttle_0 = blocks.throttle(gr.sizeof_gr_complex*1, samp_rate/zoom_ratio,True)

        ##################################################
        # Connections
        ##################################################
        self.connect((self.blocks_throttle_0, 0), (self.wxgui_fftsink2_0, 0))
        self.connect((self.low_pass_filter_0, 0), (self.blocks_throttle_0, 0))
        self.connect((self.osmosdr_source_1, 0), (self.low_pass_filter_0, 0))

    def get_ratio_chooser(self):
        return self.ratio_chooser

    def set_ratio_chooser(self, ratio_chooser):
        self.ratio_chooser = ratio_chooser
        self.set_zoom_ratio(self.ratio_chooser)
        self._ratio_chooser_chooser.set_value(self.ratio_chooser)

    def get_zoom_ratio(self):
        return self.zoom_ratio

    def set_zoom_ratio(self, zoom_ratio):
        self.zoom_ratio = zoom_ratio
        self.wxgui_fftsink2_0.set_sample_rate(self.samp_rate/self.zoom_ratio)
        self.low_pass_filter_0.set_taps(firdes.low_pass(1, self.samp_rate, self.samp_rate/self.zoom_ratio/2, 100, firdes.WIN_HAMMING, 6.76))
        self.blocks_throttle_0.set_sample_rate(self.samp_rate/self.zoom_ratio)

    def get_samp_rate(self):
        return self.samp_rate

    def set_samp_rate(self, samp_rate):
        self.samp_rate = samp_rate
        self.wxgui_fftsink2_0.set_sample_rate(self.samp_rate/self.zoom_ratio)
        self.osmosdr_source_1.set_sample_rate(self.samp_rate)
        self.low_pass_filter_0.set_taps(firdes.low_pass(1, self.samp_rate, self.samp_rate/self.zoom_ratio/2, 100, firdes.WIN_HAMMING, 6.76))
        self.blocks_throttle_0.set_sample_rate(self.samp_rate/self.zoom_ratio)

    def get_freq(self):
        return self.freq

    def set_freq(self, freq):
        self.freq = freq
        self.wxgui_fftsink2_0.set_baseband_freq(self.freq)
        self.osmosdr_source_1.set_center_freq(self.freq, 0)


def main(top_block_cls=zoom_fft, options=None):

    tb = top_block_cls()
    tb.Start(True)
    tb.Wait()


if __name__ == '__main__':
    main()


gnuradio/top_block.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
##################################################
# GNU Radio Python Flow Graph
# Title: Top Block
# Generated: Wed Sep  6 12:53:25 2017
##################################################

if __name__ == '__main__':
    import ctypes
    import sys
    if sys.platform.startswith('linux'):
        try:
            x11 = ctypes.cdll.LoadLibrary('libX11.so')
            x11.XInitThreads()
        except:
            print "Warning: failed to XInitThreads()"

from gnuradio import analog
from gnuradio import blocks
from gnuradio import eng_notation
from gnuradio import gr
from gnuradio import wxgui
from gnuradio.eng_option import eng_option
from gnuradio.filter import firdes
from gnuradio.wxgui import forms
from gnuradio.wxgui import scopesink2
from grc_gnuradio import blks2 as grc_blks2
from grc_gnuradio import wxgui as grc_wxgui
from optparse import OptionParser
import wx


class top_block(grc_wxgui.top_block_gui):

    def __init__(self):
        grc_wxgui.top_block_gui.__init__(self, title="Top Block")
        _icon_path = "/usr/share/icons/hicolor/32x32/apps/gnuradio-grc.png"
        self.SetIcon(wx.Icon(_icon_path, wx.BITMAP_TYPE_ANY))

        ##################################################
        # Variables
        ##################################################
        self.Ind_s = Ind_s = 0
        self.Am_s = Am_s = 1.25893
        self.samp_rate = samp_rate = 512000
        self.Ind = Ind = Ind_s
        self.Am_sig = Am_sig = Am_s

        ##################################################
        # Blocks
        ##################################################
        self.wxgui_scopesink2_0 = scopesink2.scope_sink_f(
        	self.GetWin(),
        	title='Scope Plot',
        	sample_rate=samp_rate,
        	v_scale=0,
        	v_offset=0,
        	t_scale=0,
        	ac_couple=False,
        	xy_mode=False,
        	num_inputs=1,
        	trig_mode=wxgui.TRIG_MODE_AUTO,
        	y_axis_label='Counts',
        )
        self.Add(self.wxgui_scopesink2_0.win)
        self.blocks_multiply_xx_0 = blocks.multiply_vff(1)
        self.blocks_multiply_const_vxx_0 = blocks.multiply_const_vff((Am_sig, ))
        self.blocks_add_xx_0 = blocks.add_vff(1)
        self.blocks_add_const_vxx_0 = blocks.add_const_vff((1, ))
        self.blks2_selector_0 = grc_blks2.selector(
        	item_size=gr.sizeof_float*1,
        	num_inputs=2,
        	num_outputs=1,
        	input_index=Ind,
        	output_index=0,
        )
        self.analog_sig_source_x_0_0 = analog.sig_source_f(samp_rate, analog.GR_COS_WAVE, 1000, 1, 0)
        self.analog_sig_source_x_0 = analog.sig_source_f(samp_rate, analog.GR_COS_WAVE, 30000, 1, 0)
        self.analog_noise_source_x_0 = analog.noise_source_f(analog.GR_GAUSSIAN, 1, 0)
        self._Ind_s_chooser = forms.drop_down(
        	parent=self.GetWin(),
        	value=self.Ind_s,
        	callback=self.set_Ind_s,
        	label='Ind_s',
        	choices=[0, 1],
        	labels=[],
        )
        self.Add(self._Ind_s_chooser)
        _Am_s_sizer = wx.BoxSizer(wx.VERTICAL)
        self._Am_s_text_box = forms.text_box(
        	parent=self.GetWin(),
        	sizer=_Am_s_sizer,
        	value=self.Am_s,
        	callback=self.set_Am_s,
        	label='Am_s',
        	converter=forms.float_converter(),
        	proportion=0,
        )
        self._Am_s_slider = forms.slider(
        	parent=self.GetWin(),
        	sizer=_Am_s_sizer,
        	value=self.Am_s,
        	callback=self.set_Am_s,
        	minimum=0,
        	maximum=50,
        	num_steps=100,
        	style=wx.SL_HORIZONTAL,
        	cast=float,
        	proportion=1,
        )
        self.Add(_Am_s_sizer)

        ##################################################
        # Connections
        ##################################################
        self.connect((self.analog_noise_source_x_0, 0), (self.blocks_add_xx_0, 1))
        self.connect((self.analog_sig_source_x_0, 0), (self.blocks_multiply_xx_0, 0))
        self.connect((self.analog_sig_source_x_0_0, 0), (self.blocks_add_const_vxx_0, 0))
        self.connect((self.blks2_selector_0, 0), (self.wxgui_scopesink2_0, 0))
        self.connect((self.blocks_add_const_vxx_0, 0), (self.blocks_multiply_xx_0, 1))
        self.connect((self.blocks_add_xx_0, 0), (self.blks2_selector_0, 0))
        self.connect((self.blocks_multiply_const_vxx_0, 0), (self.blks2_selector_0, 1))
        self.connect((self.blocks_multiply_const_vxx_0, 0), (self.blocks_add_xx_0, 0))
        self.connect((self.blocks_multiply_xx_0, 0), (self.blocks_multiply_const_vxx_0, 0))

    def get_Ind_s(self):
        return self.Ind_s

    def set_Ind_s(self, Ind_s):
        self.Ind_s = Ind_s
        self.set_Ind(self.Ind_s)
        self._Ind_s_chooser.set_value(self.Ind_s)

    def get_Am_s(self):
        return self.Am_s

    def set_Am_s(self, Am_s):
        self.Am_s = Am_s
        self.set_Am_sig(self.Am_s)
        self._Am_s_slider.set_value(self.Am_s)
        self._Am_s_text_box.set_value(self.Am_s)

    def get_samp_rate(self):
        return self.samp_rate

    def set_samp_rate(self, samp_rate):
        self.samp_rate = samp_rate
        self.wxgui_scopesink2_0.set_sample_rate(self.samp_rate)
        self.analog_sig_source_x_0_0.set_sampling_freq(self.samp_rate)
        self.analog_sig_source_x_0.set_sampling_freq(self.samp_rate)

    def get_Ind(self):
        return self.Ind

    def set_Ind(self, Ind):
        self.Ind = Ind
        self.blks2_selector_0.set_input_index(int(self.Ind))

    def get_Am_sig(self):
        return self.Am_sig

    def set_Am_sig(self, Am_sig):
        self.Am_sig = Am_sig
        self.blocks_multiply_const_vxx_0.set_k((self.Am_sig, ))


def main(top_block_cls=top_block, options=None):

    tb = top_block_cls()
    tb.Start(True)
    tb.Wait()


if __name__ == '__main__':
    main()


gnuradio/analog_wfm_rx.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
##################################################
# GNU Radio Python Flow Graph
# Title: Mono FM Radio
# Author: OZ9AEC
# Description: WFM receiver (mono only)
# Generated: Mon Sep  4 20:17:27 2017
##################################################

if __name__ == '__main__':
    import ctypes
    import sys
    if sys.platform.startswith('linux'):
        try:
            x11 = ctypes.cdll.LoadLibrary('libX11.so')
            x11.XInitThreads()
        except:
            print "Warning: failed to XInitThreads()"

from gnuradio import audio
from gnuradio import blocks
from gnuradio import eng_notation
from gnuradio import filter
from gnuradio import gr
from gnuradio import wxgui
from gnuradio.eng_option import eng_option
from gnuradio.fft import window
from gnuradio.filter import firdes
from gnuradio.wxgui import fftsink2
from gnuradio.wxgui import forms
from grc_gnuradio import wxgui as grc_wxgui
from optparse import OptionParser
import osmosdr
import time
import wx


class analog_wfm_rx(grc_wxgui.top_block_gui):

    def __init__(self):
        grc_wxgui.top_block_gui.__init__(self, title="Mono FM Radio")
        _icon_path = "/usr/share/icons/hicolor/32x32/apps/gnuradio-grc.png"
        self.SetIcon(wx.Icon(_icon_path, wx.BITMAP_TYPE_ANY))

        ##################################################
        # Variables
        ##################################################
        self.xlate_tune = xlate_tune = 0
        self.usrp_freq = usrp_freq = 101.8e6
        self.samp_rate = samp_rate = 2500e3
        self.rx_freq = rx_freq = usrp_freq+xlate_tune
        self.rf_gain = rf_gain = 7
        self.if_gain = if_gain = 7
        self.filter_taps = filter_taps = firdes.low_pass(1, samp_rate, 250000, 20000, firdes.WIN_HAMMING, 6.76)
        self.bb_gain = bb_gain = 7
        self.af_gain = af_gain = 0.3

        ##################################################
        # Blocks
        ##################################################
        _xlate_tune_sizer = wx.BoxSizer(wx.VERTICAL)
        self._xlate_tune_text_box = forms.text_box(
        	parent=self.GetWin(),
        	sizer=_xlate_tune_sizer,
        	value=self.xlate_tune,
        	callback=self.set_xlate_tune,
        	label='Fine frequency',
        	converter=forms.float_converter(),
        	proportion=0,
        )
        self._xlate_tune_slider = forms.slider(
        	parent=self.GetWin(),
        	sizer=_xlate_tune_sizer,
        	value=self.xlate_tune,
        	callback=self.set_xlate_tune,
        	minimum=-250e3,
        	maximum=250e3,
        	num_steps=500,
        	style=wx.SL_HORIZONTAL,
        	cast=float,
        	proportion=1,
        )
        self.GridAdd(_xlate_tune_sizer, 7, 0, 1, 5)
        _usrp_freq_sizer = wx.BoxSizer(wx.VERTICAL)
        self._usrp_freq_text_box = forms.text_box(
        	parent=self.GetWin(),
        	sizer=_usrp_freq_sizer,
        	value=self.usrp_freq,
        	callback=self.set_usrp_freq,
        	label='USRP frequency',
        	converter=forms.float_converter(),
        	proportion=0,
        )
        self._usrp_freq_slider = forms.slider(
        	parent=self.GetWin(),
        	sizer=_usrp_freq_sizer,
        	value=self.usrp_freq,
        	callback=self.set_usrp_freq,
        	minimum=88e6,
        	maximum=108e6,
        	num_steps=200,
        	style=wx.SL_HORIZONTAL,
        	cast=float,
        	proportion=1,
        )
        self.GridAdd(_usrp_freq_sizer, 6, 0, 1, 5)
        _rf_gain_sizer = wx.BoxSizer(wx.VERTICAL)
        self._rf_gain_text_box = forms.text_box(
        	parent=self.GetWin(),
        	sizer=_rf_gain_sizer,
        	value=self.rf_gain,
        	callback=self.set_rf_gain,
        	label='RF',
        	converter=forms.float_converter(),
        	proportion=0,
        )
        self._rf_gain_slider = forms.slider(
        	parent=self.GetWin(),
        	sizer=_rf_gain_sizer,
        	value=self.rf_gain,
        	callback=self.set_rf_gain,
        	minimum=0,
        	maximum=90,
        	num_steps=90,
        	style=wx.SL_HORIZONTAL,
        	cast=float,
        	proportion=1,
        )
        self.GridAdd(_rf_gain_sizer, 9, 2, 1, 1)
        self.nbook = self.nbook = wx.Notebook(self.GetWin(), style=wx.NB_TOP)
        self.nbook.AddPage(grc_wxgui.Panel(self.nbook), "Receiver")
        self.nbook.AddPage(grc_wxgui.Panel(self.nbook), "Audio")
        self.GridAdd(self.nbook, 0, 0, 5, 5)
        _if_gain_sizer = wx.BoxSizer(wx.VERTICAL)
        self._if_gain_text_box = forms.text_box(
        	parent=self.GetWin(),
        	sizer=_if_gain_sizer,
        	value=self.if_gain,
        	callback=self.set_if_gain,
        	label='IF',
        	converter=forms.float_converter(),
        	proportion=0,
        )
        self._if_gain_slider = forms.slider(
        	parent=self.GetWin(),
        	sizer=_if_gain_sizer,
        	value=self.if_gain,
        	callback=self.set_if_gain,
        	minimum=0,
        	maximum=60,
        	num_steps=60,
        	style=wx.SL_HORIZONTAL,
        	cast=float,
        	proportion=1,
        )
        self.GridAdd(_if_gain_sizer, 9, 1, 1, 1)
        _bb_gain_sizer = wx.BoxSizer(wx.VERTICAL)
        self._bb_gain_text_box = forms.text_box(
        	parent=self.GetWin(),
        	sizer=_bb_gain_sizer,
        	value=self.bb_gain,
        	callback=self.set_bb_gain,
        	label='BB',
        	converter=forms.float_converter(),
        	proportion=0,
        )
        self._bb_gain_slider = forms.slider(
        	parent=self.GetWin(),
        	sizer=_bb_gain_sizer,
        	value=self.bb_gain,
        	callback=self.set_bb_gain,
        	minimum=0,
        	maximum=90,
        	num_steps=90,
        	style=wx.SL_HORIZONTAL,
        	cast=float,
        	proportion=1,
        )
        self.GridAdd(_bb_gain_sizer, 9, 3, 1, 1)
        _af_gain_sizer = wx.BoxSizer(wx.VERTICAL)
        self._af_gain_text_box = forms.text_box(
        	parent=self.GetWin(),
        	sizer=_af_gain_sizer,
        	value=self.af_gain,
        	callback=self.set_af_gain,
        	label='AF',
        	converter=forms.float_converter(),
        	proportion=0,
        )
        self._af_gain_slider = forms.slider(
        	parent=self.GetWin(),
        	sizer=_af_gain_sizer,
        	value=self.af_gain,
        	callback=self.set_af_gain,
        	minimum=0,
        	maximum=1,
        	num_steps=100,
        	style=wx.SL_HORIZONTAL,
        	cast=float,
        	proportion=1,
        )
        self.GridAdd(_af_gain_sizer, 9, 0, 1, 1)
        self.xlating_fir_filter = filter.freq_xlating_fir_filter_ccc(1, (filter_taps), -xlate_tune, samp_rate)
        self._rx_freq_static_text = forms.static_text(
        	parent=self.GetWin(),
        	value=self.rx_freq,
        	callback=self.set_rx_freq,
        	label='Receive',
        	converter=forms.float_converter(),
        )
        self.GridAdd(self._rx_freq_static_text, 5, 3, 1, 1)
        self.rr_stereo_right = filter.rational_resampler_fff(
                interpolation=48,
                decimation=2500,
                taps=None,
                fractional_bw=None,
        )
        self.osmosdr_source_1 = osmosdr.source( args="numchan=" + str(1) + " " + 'airspy' )
        self.osmosdr_source_1.set_sample_rate(samp_rate)
        self.osmosdr_source_1.set_center_freq(usrp_freq, 0)
        self.osmosdr_source_1.set_freq_corr(0, 0)
        self.osmosdr_source_1.set_dc_offset_mode(0, 0)
        self.osmosdr_source_1.set_iq_balance_mode(0, 0)
        self.osmosdr_source_1.set_gain_mode(False, 0)
        self.osmosdr_source_1.set_gain(rf_gain, 0)
        self.osmosdr_source_1.set_if_gain(if_gain, 0)
        self.osmosdr_source_1.set_bb_gain(bb_gain, 0)
        self.osmosdr_source_1.set_antenna('', 0)
        self.osmosdr_source_1.set_bandwidth(0, 0)

        self.fftsink_rf = fftsink2.fft_sink_c(
        	self.nbook.GetPage(0).GetWin(),
        	baseband_freq=0,
        	y_per_div=10,
        	y_divs=10,
        	ref_level=-30,
        	ref_scale=1.0,
        	sample_rate=samp_rate/2,
        	fft_size=512,
        	fft_rate=10,
        	average=True,
        	avg_alpha=0.5,
        	title='Baseband',
        	peak_hold=False,
        )
        self.nbook.GetPage(0).Add(self.fftsink_rf.win)
        self.blocks_multiply_conjugate_cc_0 = blocks.multiply_conjugate_cc(1)
        self.blocks_delay_0 = blocks.delay(gr.sizeof_gr_complex*1, 1)
        self.blocks_complex_to_arg_0 = blocks.complex_to_arg(1)
        self.audio_sink = audio.sink(48000, '', True)
        self.af_gain_stereo_left = blocks.multiply_const_vff((af_gain, ))

        ##################################################
        # Connections
        ##################################################
        self.connect((self.af_gain_stereo_left, 0), (self.audio_sink, 0))
        self.connect((self.af_gain_stereo_left, 0), (self.audio_sink, 1))
        self.connect((self.blocks_complex_to_arg_0, 0), (self.rr_stereo_right, 0))
        self.connect((self.blocks_delay_0, 0), (self.blocks_multiply_conjugate_cc_0, 0))
        self.connect((self.blocks_multiply_conjugate_cc_0, 0), (self.blocks_complex_to_arg_0, 0))
        self.connect((self.osmosdr_source_1, 0), (self.xlating_fir_filter, 0))
        self.connect((self.rr_stereo_right, 0), (self.af_gain_stereo_left, 0))
        self.connect((self.xlating_fir_filter, 0), (self.blocks_delay_0, 0))
        self.connect((self.xlating_fir_filter, 0), (self.blocks_multiply_conjugate_cc_0, 1))
        self.connect((self.xlating_fir_filter, 0), (self.fftsink_rf, 0))

    def get_xlate_tune(self):
        return self.xlate_tune

    def set_xlate_tune(self, xlate_tune):
        self.xlate_tune = xlate_tune
        self._xlate_tune_slider.set_value(self.xlate_tune)
        self._xlate_tune_text_box.set_value(self.xlate_tune)
        self.xlating_fir_filter.set_center_freq(-self.xlate_tune)
        self.set_rx_freq(self.usrp_freq+self.xlate_tune)

    def get_usrp_freq(self):
        return self.usrp_freq

    def set_usrp_freq(self, usrp_freq):
        self.usrp_freq = usrp_freq
        self._usrp_freq_slider.set_value(self.usrp_freq)
        self._usrp_freq_text_box.set_value(self.usrp_freq)
        self.set_rx_freq(self.usrp_freq+self.xlate_tune)
        self.osmosdr_source_1.set_center_freq(self.usrp_freq, 0)

    def get_samp_rate(self):
        return self.samp_rate

    def set_samp_rate(self, samp_rate):
        self.samp_rate = samp_rate
        self.set_filter_taps(firdes.low_pass(1, self.samp_rate, 250000, 20000, firdes.WIN_HAMMING, 6.76))
        self.osmosdr_source_1.set_sample_rate(self.samp_rate)
        self.fftsink_rf.set_sample_rate(self.samp_rate/2)

    def get_rx_freq(self):
        return self.rx_freq

    def set_rx_freq(self, rx_freq):
        self.rx_freq = rx_freq
        self._rx_freq_static_text.set_value(self.rx_freq)

    def get_rf_gain(self):
        return self.rf_gain

    def set_rf_gain(self, rf_gain):
        self.rf_gain = rf_gain
        self._rf_gain_slider.set_value(self.rf_gain)
        self._rf_gain_text_box.set_value(self.rf_gain)
        self.osmosdr_source_1.set_gain(self.rf_gain, 0)

    def get_if_gain(self):
        return self.if_gain

    def set_if_gain(self, if_gain):
        self.if_gain = if_gain
        self._if_gain_slider.set_value(self.if_gain)
        self._if_gain_text_box.set_value(self.if_gain)
        self.osmosdr_source_1.set_if_gain(self.if_gain, 0)

    def get_filter_taps(self):
        return self.filter_taps

    def set_filter_taps(self, filter_taps):
        self.filter_taps = filter_taps
        self.xlating_fir_filter.set_taps((self.filter_taps))

    def get_bb_gain(self):
        return self.bb_gain

    def set_bb_gain(self, bb_gain):
        self.bb_gain = bb_gain
        self._bb_gain_slider.set_value(self.bb_gain)
        self._bb_gain_text_box.set_value(self.bb_gain)
        self.osmosdr_source_1.set_bb_gain(self.bb_gain, 0)

    def get_af_gain(self):
        return self.af_gain

    def set_af_gain(self, af_gain):
        self.af_gain = af_gain
        self._af_gain_slider.set_value(self.af_gain)
        self._af_gain_text_box.set_value(self.af_gain)
        self.af_gain_stereo_left.set_k((self.af_gain, ))


def main(top_block_cls=analog_wfm_rx, options=None):

    tb = top_block_cls()
    tb.Start(True)
    tb.Wait()


if __name__ == '__main__':
    main()
